Framework: add approach for redux data persistence

gwwar opened the pull request:
Data persistence is still a wip but I think we should update this as we go along.

cc @rralian @mtias @aduth @blowery @dmsnell @ockham

aduth commented on docs/our-approach-to-data.md:
````diff
@@ -271,3 +272,137 @@ By now, you're hopefully convinced that a global application state can enable us
+reducer( browserState, { type: 'DESERIALIZE' } )
+```
+
+The key idea is that in the reducer `SERIALIZE` should return a plain JS object. In a subtree that uses Immutable.js
````
Should we explain why it's important to return a JavaScript object?

```text
Because browser storage is only capable of storing simple JavaScript objects, the purpose of the `SERIALIZE` action type reducer handler is to return a plain object representation.
```

aduth commented on docs/our-approach-to-data.md:
````diff
@@ -271,3 +272,137 @@ By now, you're hopefully convinced that a global application state can enable us
+} );
+```
+
+`DESERIALIZE` in turn returns what the subtree expects. In a subtree that uses Immutable.js it looks like:
````
Should we explain that what is being converted is from browser storage?

```
In turn, when the store instance is initialized with the browser storage copy of state, you can convert your subtree state back to its expected format from the `DESERIALIZE` handler.
```

aduth commented on docs/our-approach-to-data.md:
````diff
@@ -271,3 +272,137 @@ By now, you're hopefully convinced that a global application state can enable us
+			return fromJS( state );
+} );
+```
+If the subtree uses plain JS objects, we simply return state.
````
Worth explaining that `SERIALIZE` and `DESERIALIZE` can be omitted if the reduced state is safe for browser storage?

```
If your reducer state is already a plain object, you can choose to omit `SERIALIZE` and `DESERIALIZE` handlers, or simply define them as returning the current state.
```

gwwar replied:
:+1: Yes, definitely worth explaining. Also, if a reducer chooses to omit handling those actions, the subtree may see future errors when data shapes change. I'll add a note.

aduth commented on docs/our-approach-to-data.md:
````diff
@@ -271,3 +272,137 @@ By now, you're hopefully convinced that a global application state can enable us
+```
+If the subtree uses plain JS objects, we simply return state.
+
+You may have noticed that if we leave `DESERIALIZE` as is, we will run into problems with data shapes changing over time.
````
Have I noticed this? It's not _that_ obvious :smile:

aduth commented on docs/our-approach-to-data.md:
```diff
@@ -271,3 +272,137 @@ By now, you're hopefully convinced that a global application state can enable us
+
+You may have noticed that if we leave `DESERIALIZE` as is, we will run into problems with data shapes changing over time.
+
+To work around this, we add a json schema `schema.js` to match each reducer. This schema describes what 
```
As another developer, I may be totally unfamiliar with what JSON Schema is. Could you explain it briefly and link to useful guides for getting started?

aduth commented on docs/our-approach-to-data.md:
````diff
@@ -271,3 +272,137 @@ By now, you're hopefully convinced that a global application state can enable us
+} );
+```
+
+All subtrees are highly encouraged to implement `SERIALIZE` and `DESERIALIZE` to avoid errors when data shape changes. A 
````
"All subtrees implementing" is slightly confusing wording. Consider:

```
You are encouraged to implement `SERIALIZE` and `DESERIALIZE` in your reducers...
```

aduth commented on docs/our-approach-to-data.md:
````diff
@@ -271,3 +272,137 @@ By now, you're hopefully convinced that a global application state can enable us
+```
+
+All subtrees are highly encouraged to implement `SERIALIZE` and `DESERIALIZE` to avoid errors when data shape changes. A 
+json schema should be added as well if the subtree chooses to persist state.
````
Proper capitalization should be "JSON Schema"

aduth commented:
The headings are a bit confusing to the flow of reading:

- Persistence
 - Problem
 - Problem
 - Solution
 - Solution
- Not persisting

I might expect it to be structured as:

- Persistence
  - Problems
    - Problem
      - Solution
    - Problem
      - Solution
  - Not persisting

gwwar commented:
Thanks for the :eyes: @aduth! Great suggestions. I hopefully addressed most of the issues in 7057d5a I opted to not go more than 2 deep with headers, because the styling looks a bit odd.

rralian commented:
I think this looks great... thanks for writing it up! Looks good to :ship: to me.

gwwar commented:
Thanks for the reviews! Going to squash commits.
